zkevm-circuits/src/util/cell_manager_strategy.rs

```diff
@@ -0,0 +1,244 @@
+    ) -> CMFixedWidthStrategy {
+        CMFixedWidthStrategy {
+            advices,
+            height_offset,
+            next: HashMap::default(),
+            perm_substitution: false,
+        }
+    }
+
+    pub fn with_perm_substitution(mut self) -> Self {
+        self.perm_substitution = true;
+
+        self
+    }
+
+    fn get_next(&self, cell_type: &CellType) -> (usize, usize) {
+        *self.next.get(cell_type).unwrap_or(&(0, 0))
+    }
+
+    fn set_next(&mut self, cell_type: &CellType, column_idx: usize, row: usize) {
+        self.next.insert(*cell_type, (column_idx, row));
+    }
+
+    fn cells_used(&self, cell_type: &CellType, columns: &CellManagerColumns) -> usize {
+        let (next_column_idx, next_row) = self.get_next(cell_type);
+        let current_row = if next_column_idx == 0 {
+            if next_row == 0 {
+                return 0;
+            }
+
+            next_row - 1
+        } else {
+            next_row
+        };
+
+        let filled_rows_cells = if current_row == 0 {
+            0
+        } else {
+            (current_row - 1) * columns.get_cell_type_width(*cell_type)
+        };
+
+        filled_rows_cells + next_column_idx
+    }
+}
+
+impl CellManagerStrategy for CMFixedWidthStrategy {
+    fn on_creation(&mut self, columns: &mut CellManagerColumns) {
+        for (cell_type, advices) in self.advices.0.iter() {
+            for column in advices.iter() {
+                columns.add_column(*cell_type, *column)
+            }
+        }
+    }
+
+    fn query_cell<F: Field>(
+        &mut self,
+        columns: &mut CellManagerColumns,
+        meta: &mut ConstraintSystem<F>,
+        cell_type: CellType,
+    ) -> Cell<F> {
+        let (mut column_idx, mut row) = self.get_next(&cell_type);
+        if self.perm_substitution && cell_type == CellType::StoragePhase1 {
+            let (_, row_perm) = self.get_next(&CellType::StoragePermutation);
+            if row_perm < row {
+                return self.query_cell(columns, meta, CellType::StoragePermutation);
+            }
+        }
+
+        let column = columns
+            .get_column(cell_type, column_idx)
+            .expect("column not found");
+
+        let cell = Cell::new_from_cs(meta, column.advice, column.idx, self.height_offset + row);
+
+        column_idx += 1;
+        if column_idx >= columns.get_cell_type_width(cell_type) {
+            column_idx = 0;
+            row += 1;
+        }
+
+        self.set_next(&cell_type, column_idx, row);
+
+        cell
+    }
+
+    fn get_height(&self) -> usize {
+        self.next
+            .keys()
+            .map(|cell_type| {
+                let next = self.get_next(cell_type);
+                if next.0 == 0 {
+                    next.1
+                } else {
+                    next.1 + 1
+                }
+            })
+            .max()
+            .unwrap_or(0)
+    }
+
+    type Stats = BTreeMap<CellType, (usize, usize, usize)>;
```

Comment:
I suggest moving this to the line after `impl CellManagerStrategy for CMFixedWidthStrategy {` to be consistent with the codebase.

Reply:
Done